LSM: Avoid sorting TableMutable when updated in sorted order.

kprotty opened the pull request:
Implements a wrapper around `[]Value`, `HashMap(u24)`, and `[]u8` for TableMutable. Values are appended to the array, their indexes are added to the HashMap, and there's a bitset to track which ones were removed. When iterating, removed Values are ignored using the bitset and we track if the values were inserted in sorted order. If so, we can skip `std.sort.sort`.

## Pre-merge checklist

Performance:

* [x] Compare `zig build benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    1223 batches in 133.25 s
    load offered = 1000000 tx/s
    load accepted = 75047 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 27 ms
    batch latency p20 = 28 ms
    batch latency p30 = 29 ms
    batch latency p40 = 29 ms
    batch latency p50 = 30 ms
    batch latency p60 = 31 ms
    batch latency p70 = 32 ms
    batch latency p80 = 33 ms
    batch latency p90 = 34 ms
    batch latency p100 = 5404 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 4066 ms
    transfer latency p20 = 10537 ms
    transfer latency p30 = 20729 ms
    transfer latency p40 = 33777 ms
    transfer latency p50 = 48780 ms
    transfer latency p60 = 63997 ms
    transfer latency p70 = 74013 ms
    transfer latency p80 = 89395 ms
    transfer latency p90 = 104911 ms
    transfer latency p100 = 123261 ms
    
    # benchmark results after
    1223 batches in 123.42 s
    load offered = 1000000 tx/s
    load accepted = 81021 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 26 ms
    batch latency p20 = 27 ms
    batch latency p30 = 28 ms
    batch latency p40 = 28 ms
    batch latency p50 = 29 ms
    batch latency p60 = 30 ms
    batch latency p70 = 30 ms
    batch latency p80 = 31 ms
    batch latency p90 = 33 ms
    batch latency p100 = 5406 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 3215 ms
    transfer latency p20 = 9149 ms
    transfer latency p30 = 18426 ms
    transfer latency p40 = 30364 ms
    transfer latency p50 = 44402 ms
    transfer latency p60 = 58471 ms
    transfer latency p70 = 67372 ms
    transfer latency p80 = 81779 ms
    transfer latency p90 = 96208 ms
    transfer latency p100 = 113436 ms
    ```

kprotty committed:
lsm: avoid sorted when put in sorted order

Implements a wrapper around `[]Value`, `HashMap(u32)`, and `[]u8`
for TableMutable. Values are appended to the array, their indexes are added
to the HashMap, and there's a bitset to track which ones were removed.
When iterating, removed Values are ignored using the bitset and we track
if the values were inserted in sorted order. If so, we can skip `std.sort.sort`.

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -174,10 +297,16 @@ pub fn TableMutableType(comptime Table: type, comptime tree_name: [:0]const u8)
+                if (i > 0 and sorted) {
+                    const prev_key = key_from_value(&values_max[i - 1]);
+                    sorted = compare_keys(prev_key, key_from_value(value)) != .gt;
+                }
```
This is expensive to do in the data plane, since we're branching on `sorted` and comparing all keys.

The idea is that by this point, TableMutable should already know whether it needs to do any last minute sort finalization, or else if all keys are already sorted, and if so, it should be aiming to do a single memcpy here.

kprotty replied:
The comparison of all keys to check if it's sorted (stops comparing when it's not sorted) either happens here or during `upsert/insert_no_clobber`. We can't do a memcpy as some Values may be deleted from `fetch_remove` and shouldn't get copied into `values_max` (the `active` bitset is used to detect this). It's also already traversed for updating the values_cache.

jorangreef replied:
> The comparison of all keys to check if it's sorted (stops comparing when it's not sorted) either happens here or during upsert/insert_no_clobber.

Yes, however, doing it during upsert/insert would be more incremental.

> We can't do a memcpy as some Values may be deleted from fetch_remove and shouldn't get copied into values_max (the active bitset is used to detect this).

I mean that we should *aim* to do a single memcpy, i.e. we can iterate the bits quickly to find the biggest memcpy we can do. Then skip. Then big memcpy etc.

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -127,23 +254,19 @@ pub fn TableMutableType(comptime Table: type, comptime tree_name: [:0]const u8)
-                    const upsert = table.values.getOrPutAssumeCapacity(value.*);
-                    upsert.key_ptr.* = tombstone_from_key(key_from_value(value));
+                    // Overwrite the old key and value if any.
+                    table.values.upsert(value).* = tombstone_from_key(key_from_value(value));
```
Nice API improvement here with `upsert()`.

jorangreef replied:
We could improve the comment, since this not only overwrites the old key if any, but will otherwise also insert.

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -42,7 +39,142 @@ pub fn TableMutableType(comptime Table: type, comptime tree_name: [:0]const u8)
+
+            pub fn upsert(map: *ValuesMap, value: *const Value) *Value {
+                const ctx = IndexMapContext{ .values = map.values, .search_key = key_from_value(value) };
+                const result = map.index_map.getOrPutAssumeCapacityContext(ctx.search_index, ctx);
```
Is this correct? Always passing `maxInt(u32)` (`ctx.search_index`) as the key to `getOrPutAssumeCapacityContext()`?

We need a fuzz test for this structure to make sure we're covering all the functionality we expect from it.

kprotty replied:
`IndexMap` deals with indexes instead of Keys. `maxInt(u32)` is used as a dummy index to represent the key being passed in for lookup instead of the Key of a Value already in the array. `IndexMapContext` does the appropriate conversion for the hashmap.

kprotty replied:
This was replaced with a custom map implementation.

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -42,7 +39,142 @@ pub fn TableMutableType(comptime Table: type, comptime tree_name: [:0]const u8)
+                const kv = map.index_map.fetchRemoveContext(ctx.search_index, ctx) orelse return null;
+                const index = kv.key;
+
+                map.values.active[index / 8] &= ~(@as(u8, 1) << @intCast(u3, index % 8));
```
```zig
[index / 8]
```

As per TigerStyle, let's always use `@divExact()` etc. to indicate what we intend precisely with respect to flooring.

jorangreef replied:
Also, is there anything we expect of the existing value for `active[index / 8]` here?

Or do we intend to unilaterally `AND` it regardless of the existing value?

jorangreef replied:
Is there anything we need to still do with respect to `map.active`?

kprotty replied:
Yes, we expect its bit to already be set. Ill add that.

kprotty committed:
lsm: optimize TableMutable operations

Implements a SwissMap with F14 probing instead of using std containers.
This results is less memory overhead per Value and faster collision resolution:
* old per-Value overhead: `(u8 + u32) * 2 + u1` = `u81`
* new per-Value overhead: `(u8 + u24) * 2 + u1` = `u65`

This also avoids `insert_no_clobber` having to hash and lookup again after
a `fetch_remove`.

kprotty committed:
lsm: fix TableMutable find() usage.

kprotty committed:
lsm: maintain TableMutable sorted invariant via inserts

kprotty commented:
superseded by #784
